url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/compiler/parser/XTokenizer.java

```diff
@@ -0,0 +1,385 @@
+
+import javax.tools.DiagnosticListener;
+
+import xscript.compiler.XDiagnostic;
+import xscript.compiler.XFileReader;
+import xscript.compiler.XPosition;
+import xscript.compiler.treemaker.XKeyword;
+
+public class XTokenizer implements Iterator<XToken> {
+
+	private static final int EOF = -1;
+	
+	private XFileReader fileReader;
+	private DiagnosticListener<String> diagnosticListener;
+	
+	private XPosition start;
+	private XToken next;
+	private XToken token;
+	private int c;
+	private int spaces;
+	private int newLines;
+	private String doc=null;
+	private List<XDiagnostic> diagnostics = new ArrayList<XDiagnostic>();
+	
+	private List<XToken> marks = new ArrayList<XToken>();
+	private List<Integer> pos = new ArrayList<Integer>();
+	private int acPos;
+	
+	public XTokenizer(XFileReader fileReader, DiagnosticListener<String> diagnosticListener) {
+		this.fileReader = fileReader;
+		this.diagnosticListener = diagnosticListener;
+		nextChar();
+	}
+	
+	public XToken getToken(){
+		return token;
+	}
+	
+	@Override
+	public XToken next(){
+		if(next==null){
+			hasNext();
+		}
+		token=next;
+		next=null;
+		return token;
+	}
+
+	@Override
+	public boolean hasNext() {
+		if(next==null){
+			if(token!=null && token.kind==XTokenKind.EOF){
+				next = token;
+				return false;
+			}else{
+				next = tryGetNext();
+			}
+		}
+		return next.kind!=XTokenKind.EOF;
+	}
+
+	private void nextChar(){
+		c = fileReader.read();
+	}
+	
+	private XToken tryGetNext(){
+		if(!pos.isEmpty() && marks.size()>acPos){
+			return marks.get(acPos++);
+		}else if(pos.isEmpty() && !marks.isEmpty()){
+			return marks.remove(0);
+		}
+		XToken token = tryGetNext2();
+		if(!pos.isEmpty()){
+			marks.add(token);
+			acPos++;
+		}
+		return token;
+	}
+	
+	private XToken tryGetNext2(){
+		spaces = 0;
+		newLines = 0;
+		while(true){
+			while(c==' '||c=='\t'||c=='\r'||c=='\n'){
+				if(c==' '||c=='\t'){
+					spaces++;
+				}else if(c=='\n'){
+					newLines++;
+				}
+				nextChar();
+			}
+			start = fileReader.getPosition();
+			if(c==EOF){
+				return makeToken(XTokenKind.EOF, null);
+			}else if(c=='"' || c=='\''){
+				return makeString(c=='\'');
+			}else if(charOkForRadix(10, c)){
+				return makeNumber(false);
+			}else if(Character.isAlphabetic((char)c) || c=='_'){
+				return makeIdent();
+			}else if(c=='/'){
+				nextChar();
+				if(c=='/' || c=='*'){
+					readLineComment(c=='*');
+				}else{
+					return makeToken(XTokenKind.KEYWORD, XKeyword.DIV);
+				}
+			}else{
+				if(c=='.'){
+					nextChar();
+					if(charOkForRadix(10, c)){
+						return makeNumber(true);
+					}
+					return makeToken(XTokenKind.KEYWORD, XKeyword.DOT);
+				}else{
+					XKeyword kw = XKeyword.getKeyword(Character.toString((char)c));
+					if(kw!=null){
+						nextChar();
+						return makeToken(XTokenKind.KEYWORD, kw);
+					}else{
+						addDiagnostic2("char.unknown", (char)c);
+						nextChar();
+					}
+				}
+			}
+		}
+	}
+	
+	private XToken makeString(boolean makeChar){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 21. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 138-176, lines 177-183 into separate methods.